add handbrake rockon

mikey0000 opened the pull request:
> To ease and accelerate the PR submission, please use the template below to provide all requested information.
> You can find guidelines on which docker image to use in the [Rockstor's documentation](http://rockstor.com/docs/contribute_rockons.html#which-docker-image-should-i-use), 
> as well as examples of proper formatting in other rock-ons ([here](https://github.com/rockstor/rockon-registry/blob/master/teamspeak3.json)
> and [here](https://github.com/rockstor/rockon-registry/blob/master/nextcloud-official.json))



Fixes # .

### General information on project
This pull request proposes to add a new rock-on for the following project:
- name: Handbrake
- website:
- description: 

### Information on docker image
- docker image: <link to image page on docker hub>
- is an official docker image available for this project?: 


### Checklist
- [x] Passes [JSONlint](https://jsonlint.com) validation
- [x] Entry added to `root.json` in _alphabetical_ order (for new rock-on only)
- [x] `"description"` object lists and links to the docker image used
- [x] `"description"` object provides information on the image's particularities (advantage over another existing rock-on for the same project, for instance)
- [x] `"website"` object links to project's main website

mikey0000 commented:
I'll work my way through the checklist but I have this running on my rockstor, its pretty awesome.

phillxnet commented:
@mikey0000 Thanks for getting started on this. Would be a 'nice to have' and well done getting it up and running already.

FroggyFlox commented:
Thanks @mikey0000 !

It is indeed a very good addition to make, and has actually been requested on our forum in the past. I'm linking two of these below for reference so that we can update those threads as needed.
- https://forum.rockstor.com/t/request-rockon-handbrake/1963
- https://forum.rockstor.com/t/handbrake-running-in-docker-container/6377

Thanks 


> I'll work my way through the checklist

Thank you for that! In case this helps, feel free to use other "up-to-date" rock-ons such as [nextcloud-official.json](https://github.com/rockstor/rockon-registry/blob/master/nextcloud-official.json) or [ecodms18-09.json](https://github.com/rockstor/rockon-registry/blob/master/ecodms18-09.json) as reference for the style formatting of the `description` and `more_info` fields.

Thanks again!

mikey0000 commented:
Hi Guys, updated and checklist checked, Minor change that I've added the VNC port as well since that is pretty nice to have in there by default. Please review and let me know if I've missed anything.

phillxnet commented:
@mikey0000 Thanks for the permissions fixes. Been meaning to do this for ages. Although our deployment system sorts it on the fly it's best by way of example if the repo at least has consistency in this regard. It's generally better all around to try and avoid 'mission creep' within a pr, especially if we have to revert it's changes, but now worries; and will be nice to have this niggle finally sorted.

Nice find re the closing paragraph, and same goes for the 'mission creep' but again, nice to have this found and fixed.

Thanks for the various improvement, hopefully we can get this reviewed and in pretty soon.

Incidentally, you can tick, via GitHub web interface, the checklist: no need to edit manually. They then shows up as ticks and an associated progress meter is drawn in the pull requests overview.

mikey0000 commented:
Ah was wondering that, thanks. Yeah the mission creep was a oh I can see this needs changing, as I was working. Papercuts I'd call them. Wouldn't have been hard to turn into a separate pr, totally understand. I at least made the chmod a separate commit for that reason.

Anyway, thanks for the quick responses.

25/01/2020 11:06:49 pm Philip Guyton <notifications@github.com>:

> 
> @mikey0000 [https://github.com/mikey0000] Thanks for the permissions fixes. Been meaning to do this for ages. Although our deployment system sorts it on the fly it's best by way of example if the repo at least has consistency in this regard. It's generally better all around to try and avoid 'mission creep' within a pr, especially if we have to revert it's changes, but now worries; and will be nice to have this niggle finally sorted.
> 
> Nice find re the closing paragraph, and same goes for the 'mission creep' but again, nice to have this found and fixed.
> 
> Thanks for the various improvement, hopefully we can get this reviewed and in pretty soon.
> 
> Incidentally, you can tick, via GitHub web interface, the checklist: no need to edit manually. They then shows up as ticks and an associated progress meter is drawn in the pull requests overview.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [https://github.com/rockstor/rockon-registry/pull/210?email_source=notifications&email_token=AALDOPOZKVJNQFBFTRHFMF3Q7QFINA5CNFSM4KK4DMP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ4ZLBY#issuecomment-578393479] , or unsubscribe [https://github.com/notifications/unsubscribe-auth/AALDOPJCBFMCZGLBJV5UOTDQ7QFINANCNFSM4KK4DMPQ] . [https://github.com/notifications/beacon/AALDOPJHUYQD7QJ7ZCGWQX3Q7QFINA5CNFSM4KK4DMP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ4ZLBY.gif]
>

FroggyFlox commented on handbrake.json:
```diff
@@ -0,0 +1,64 @@
+        "icon": "https://upload.wikimedia.org/wikipedia/commons/d/d9/HandBrake_Icon.png",
+        "more_info": "<p>Additional volumes can be added to provide extra watch folders.</p><p>Documentation on how to access using VNC, adding security and configuring extra watch folders can be found here https://github.com/jlesage/docker-handbrake#accessing-the-gui</p>",
+        "ui": {
+            "slug": "gui"
```
In my test the `/gui` slug led me to a "404 page not found". I couldn't see reference to this slug in the image's documentation, so I removed it and worked perfectly for me. I thus believe we should remove it:

```suggestion
            "slug": ""
```

FroggyFlox commented on handbrake.json:
```diff
@@ -0,0 +1,64 @@
+            "slug": "gui"
+        },
+        "volume_add_support": true,
+        "website": "https://hub.docker.com/r/jlesage/handbrake/",
```
Link to project's website itself rather than docker image:

```suggestion
        "website": "https://handbrake.fr/",
```

FroggyFlox commented on handbrake.json:
```diff
@@ -0,0 +1,64 @@
+        },
+        "description": "Handbrake for converting video from nearly any format to a selection of modern, widely supported codecs. <p>This docker image makes handbrake accessible through your browser <strong>HOSTNAME:5900</strong> or VNC, see <em>more info</em> for details.</p><p>Based on the jlesage/handbrake docker image: <a href='https://hub.docker.com/r/jlesage/handbrake' target='_blank'>https: ...
+        "icon": "https://upload.wikimedia.org/wikipedia/commons/d/d9/HandBrake_Icon.png",
+        "more_info": "<p>Additional volumes can be added to provide extra watch folders.</p><p>Documentation on how to access using VNC, adding security and configuring extra watch folders can be found here https://github.com/jlesage/docker-handbrake#accessing-the-gui</p>",
```
Thanks for linking to the image's documentation, especially given how detailed it is.
I also like the note on adding volumes--I'll simply further build open this and link to Rockstor's documentation on how to do so:
```suggestion
        "more_info": "<p>Additional volumes can be added to provide extra watch folders (see <a href='http://rockstor.com/docs/docker-based-rock-ons/overview.html#add-storage' target='_blank'>Rockstor's documentation</a> for instructions on how to do so).</p><p>Documentation on how to access using VNC, adding security and configuring extra watch folders can be found here: <a href='https://github.com/jlesage/docker-handbrake#accessing-the-gui' target='_blank'>https://github.com/jlesage/docker-handbrake#accessing-the-gui</a></p>",
```

FroggyFlox replied:
@mikey0000 , could you include that one as well?
I'd prefer to see the link to Rockstor's doc being included there as well.

FroggyFlox commented on handbrake.json:
```diff
@@ -0,0 +1,64 @@
+                }
+            }
+        },
+        "description": "Handbrake for converting video from nearly any format to a selection of modern, widely supported codecs. <p>This docker image makes handbrake accessible through your browser <strong>HOSTNAME:5900</strong> or VNC, see <em>more info</em> for details.</p><p>Based on the jlesage/handbrake docker image: <a href='https://hub.docker.com/r/jlesage/handbrake' target='_blank'>https: ...
```
As the HOSTNAME:port depends on the user's install and is taken care of by Rockstor, maybe we should simply refer to the UI button (in addition to the VNC feature, of course):

```suggestion
        "description": "Handbrake for converting video from nearly any format to a selection of modern, widely supported codecs. <p>This docker image makes HandBrake accessible through your browser (<em>HandBrake UI</em> button) or VNC, see <em>more info</em> for details.</p><p>Based on the jlesage/handbrake docker image: <a href='https://hub.docker.com/r/jlesage/handbrake' target='_blank'>https://hub.docker.com/r/jlesage/handbrake</a>.</p>",
```

FroggyFlox commented:
Hi again, @mikey0000 ,

As mentioned in my review, I had another point I wanted to discuss: support for hardware acceleration via Intel QuickSync.

This point was brought up in one of the forum links I included in my earlier message, and I was thus thinking we may benefit from adding it to this rock-on. We can indeed support this by using the `"devices"` object. As this is optional, users without a compatible CPU can simply leave the corresponding field empty at installation and not be impacted. Those with compatible CPU, though, should be able to take advantage of their hardware.
What do you think?

It would simply require adding the device as we already do in the EmbyServer rock-on, for instance:
https://github.com/rockstor/rockon-registry/blob/de2679906e4972468773d0bbc036fdf60775b96e/embyserver.json#L13-L18

mikey0000 commented:
Yeah I'll add that. Was also looking at the nvenc stuff too. But I'd have to put my gpu into my server to try it, not sure what kind of gain or how many actually want it.

26/01/2020 5:49:43 am FroggyFlox <notifications@github.com>:

> 
> Hi again, @mikey0000 [https://github.com/mikey0000] ,
> 
> As mentioned in my review, I had another point I wanted to discuss: support for hardware acceleration via Intel QuickSync.
> 
> This point was brought up in one of the forum links I included in my earlier message, and I was thus thinking we may benefit from adding it to this rock-on. We can indeed support this by using the "devices" object. As this is optional, users without a compatible CPU can simply leave the corresponding field empty at installation and not be impacted. Those with compatible CPU, though, should be able to take advantage of their hardware.
> What do you think?
> 
> It would simply require adding the device as we already do in the EmbyServer rock-on, for instance:
> https://github.com/rockstor/rockon-registry/blob/de2679906e4972468773d0bbc036fdf60775b96e/embyserver.json#L13-L18 [https://github.com/rockstor/rockon-registry/blob/de2679906e4972468773d0bbc036fdf60775b96e/embyserver.json#L13-L18]
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [https://github.com/rockstor/rockon-registry/pull/210?email_source=notifications&email_token=AALDOPMLBCZWRUV35LKX7CLQ7RUPJA5CNFSM4KK4DMP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ5AKLI#issuecomment-578422061] , or unsubscribe [https://github.com/notifications/unsubscribe-auth/AALDOPITDYGVJ5QG6CVT5ULQ7RUPJANCNFSM4KK4DMPQ] . [https://github.com/notifications/beacon/AALDOPL6INEGC5RYIIWVZC3Q7RUPJA5CNFSM4KK4DMP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ5AKLI.gif]
>

FroggyFlox commented:
> Was also looking at the nvenc stuff too. But I'd have to put my gpu into my server to try it, not sure what kind of gain or how many actually want it

I agree with you about that: not sure it's the trouble. Plus, from memory, one would also need to run docker with the nvidia runtime which is not currently doable from Rockstor's UI so I don't think this is something we should try to support from the rock-on json. Note that we recently changed how we deal with the docker daemon (see https://github.com/rockstor/rockstor-core/pull/2090), so this level of configuration is now more easily thinkable in the future, though (see https://github.com/rockstor/rockstor-core/issues/2088#issuecomment-555115579).

Thanks again for your contributions!

mikey0000 commented:
@FroggyFlox I added the devices json and re-installed handbrake and didn't see the devices option in the setup flow? Any ideas help. My Server CPU is a AMD 3400G so has integrated graphics so I can test this out.

FroggyFlox commented:
> @FroggyFlox I added the devices json and re-installed handbrake and didn't see the devices option in the setup flow? Any ideas help.

Thanks a lot for implementing this. I do see the option displayed during the rock-on install wizard. As to why you do not, I can think of a couple things:
1. Have you un-installed the handbrake rock-on (if installed), and then clicked on the "Update" button in the rock-ons page? This is required to refresh the database from the newly-updated `handbrake.json` file.
2. What version of rockstor was this tested on? Source or rpm install? In any case, the support for `devices` was added in Rockstor 3.9.2-39, so you need to have this version minimum to see it. (see [Rockstor release changelogs](https://forum.rockstor.com/t/3-9-2-stable-channel-changelog/5741/2?u=flox)

Thanks again for your work, it does look good after a quick, but I'll need to double check myself later tonight.

Hope this helps,

mikey0000 commented:
Just needed to refresh 🤦‍♂

mikey0000 commented:
anything left to resolve here?

FroggyFlox commented on root.json:
```diff
@@ -20,6 +20,7 @@
     "Gogs": "gogs.json",
     "Gollum": "gollum.json",
     "HAProxy-Letsencrypt": "haproxy-letsencrypt.json",
+    "Handbrake": "handbrake.json",
```
Minor typo. I'm unsure whether or not it would actually confuse the parsing of information, but in the doubt I'd rather have it consistent with the rock-on name from the JSON.

```suggestion
    "HandBrake": "handbrake.json",
```

phillxnet commented:
@mikey0000 and @FroggyFlox I'll do a final test on this soon and hopefully get it merged and published directly thereafter.

mikey0000 commented:
@FroggyFlox I've updated the more info, I think this represents what you were asking, please review and let me know otherwise 🥇

phillxnet commented:
@mikey0000 I think a force push has confused me a little on this pull request so me and @FroggyFlox  have had a little look and think it best we merge as is as it's now a really long tail. Plus it's 99% there now. Apologies for that, we just want to get it right before publish so I'm going to merge as-is and hopefully @FroggyFlox can then more easily do a tidy as he sees fit on another pull request before we actually publish.

Thanks to both of you for getting this done. Merging as is to simplify applying the finishing touches before publishing. I'll update this pr to indicate when it's actually published.
